src/gp/src/GPMSA.C

```diff
@@ -497,6 +693,195 @@ GPMSAFactory<V, M>::addSimulations(
+
+  const unsigned int num_discrepancy_bases = m_discrepancyBases.size();
+  const unsigned int Brows = m_numExperiments * numOutputs;
+  const unsigned int Bcols =
+    m_numExperiments * (num_discrepancy_bases + num_svd_terms);
+
+  const Map B_row_map(Brows, 0, comm);
+
+  m_BMatrix.reset
+    (new M(m_simulationOutputs[0]->env(), B_row_map, Bcols));
+
+  const unsigned int Wyrows = m_numExperiments * numOutputs;
+
+  const Map Wy_row_map(Wyrows, 0, comm);
+
+  m_observationErrorMatrix.reset
+    (new M(m_simulationOutputs[0]->env(), Wy_row_map, Wyrows));
+
+  M& B = *m_BMatrix;
+  M& Wy = *m_observationErrorMatrix;
+
+  for (unsigned int ex = 0; ex != m_numExperiments; ++ex)
+    {
+      const M & D_i = m_discrepancyMatrices[ex];
+
+      // For the multivariate case, the bases K_eta computed from
+      // simulator outputs are the same as the bases K_i which apply
+      // to quantities of interest, because simulator outputs are QoIs
+      // alone.
+      //
+      // FIXME - we need to interpolate K_i in the functional case.
+
+      for (unsigned int outi = 0; outi != numOutputs; ++outi)
+        {
+          unsigned int i = ex*numOutputs+outi;
+          for (unsigned int outj = 0; outj != num_discrepancy_bases; ++outj)
+            {
+              unsigned int j = ex + m_numExperiments * outj;
+
+              B(i,j) = D_i(outi,outj);
+            }
+
+          for (unsigned int outj = 0; outj != num_svd_terms; ++outj)
+            {
+              unsigned int j = ex +
+                m_numExperiments * (num_discrepancy_bases + outj);
+
+              B(i,j) = (*m_TruncatedSVD_simulationOutputs)(outi,outj);
+            }
+
+          for (unsigned int outj = 0; outj != numOutputs; ++outj)
+            {
+              // No fancy perturbation here
+              unsigned int j = ex*numOutputs+outj;
+
+              Wy(i,j) = m_observationErrorMatrices[ex](outi,outj);
+            }
+        }
+    }
+
+  M BT_Wy_B (B.transpose() * Wy * B);
+
+  // Adding a "small ridge" to make sure this is invertible, as on
+  // p.577 - using 1e-4 from discussion notes.
+  for (unsigned int i=0; i != Brows; ++i)
+    BT_Wy_B(i,i) += 1.e-4;
+
+  BT_Wy_B_inv.reset(new M(BT_Wy_B.inverse()));
+
+  this->setUpHyperpriors();
+
+  this->m_constructedGP = true;
+  this->gpmsaEmulator.reset
+    (new GPMSAEmulator<V, M>(
+      this->prior().imageSet(),
+      this->m_scenarioSpace,
+      this->m_parameterSpace,
+      this->m_simulationOutputSpace,
+      this->m_experimentOutputSpace,
+      this->m_numSimulations,
+      this->m_numExperiments,
+      this->m_simulationScenarios,
+      this->m_simulationParameters,
+      this->m_simulationOutputs,
+      this->m_experimentScenarios,
+      this->m_experimentOutputs,
+      this->m_discrepancyBases,
+      this->m_observationErrorMatrices,
+      *(this->m_experimentErrors),
+      *(this->m_totalPrior),
+      *this->residual,
+      *this->BT_Wy_B_inv,
+      *this->KT_K_inv));
```

Comment:
Note to self, the GPMSAEmulator ABI changed (new ctor arguments added), but it's encapsulated by the GPMSAFactory.  Furthermore, although these things are not needed in the scalar case, they are still calculated but they are just ignored.